controller/object_store_controller.go

```diff
@@ -0,0 +1,1250 @@
+	if store.Spec.TargetState == longhorn.ObjectStoreStateStopped {
+		logrus.Infof("object store %v is now stopping", store.Name)
+		store.Status.State = longhorn.ObjectStoreStateStopping
+		return nil
+	}
+
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find deployment %v", store.Name)
+	} else if err = osc.checkDeployment(dpl, store); err != nil {
+		logrus.Errorf("Object Store running but deployment not ready")
+		return err
+	}
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find service %v", store.Name)
+	}
+
+	vol, err := osc.ds.GetVolume(store.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find volume %v", store.Name)
+	} else if err = osc.checkVolume(vol); err != nil {
+		logrus.Errorf("Object Store running but Volume not ready")
+		return err
+	}
+
+	pv, err := osc.ds.GetPersistentVolume(store.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find PV %v", store.Name)
+	} else if err = osc.checkPV(pv); err != nil {
+		logrus.Errorf("Object Store running but PV not ready")
+		return err
+	}
+
+	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, store.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find pvc %v", store.Name)
+	} else if err = osc.checkPVC(pvc); err != nil {
+		logrus.Errorf("Object Store running but PVC not bound")
+		return err
+	}
+
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopping(store *longhorn.ObjectStore) (err error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+		return errors.Wrap(err, "failed find Deployment to stop")
+	} else if (*dpl).Spec.Replicas != nil && *((*dpl).Spec.Replicas) != 0 {
+		(*dpl).Spec.Replicas = util.Int32Ptr(0)
+		_, err = osc.ds.UpdateDeployment(dpl)
+		return err
+	} else if dpl.Status.AvailableReplicas > 0 {
+		return nil // wait for shutdown
+	}
+
+	logrus.Infof("object store %v is now stopped", store.Name)
+	store.Status.State = longhorn.ObjectStoreStateStopped
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopped(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateRunning {
+		logrus.Infof("object store %v is now starting", store.Name)
+		store.Status.State = longhorn.ObjectStoreStateStarting
+		return nil
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
+	// The resources are created in the order:
+	// Volume -> PV -> PVC -> Deployment -> Service
+	// so we tear them down in reverse:
+	// Service -> Deployment -> PVC -> PV -> Volume
+	// Once that is done we can remove the finalizer, which allows the K8s API to
+	// remove the object store resource as well.
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err == nil {
+		return osc.ds.DeleteService(osc.namespace, store.Name)
```

Comment:
This means that if the osc.ds.DeleteService() returns nil, we will return nil and never requeue the ObjectStore object again until 1 hour resync period hit. Could you take a look at the suggesting modification change below